src/collective/fingerpointing/browser/logview.py

```diff
@@ -1,63 +1,78 @@
 # -*- coding: utf-8 -*-
-from collective.fingerpointing.logger import logfile as LOGFILE
-from glob import glob
+from collective.fingerpointing.logger import log_info
+from collective.fingerpointing.utils import all_audit_log_lines
 from Products.Five.browser import BrowserView
-from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
+
+
+DEFAULT_SIZE = 100
 
 
 class LogView(BrowserView):
 
-    index = ViewPageTemplateFile('logview.pt')
+    @property
+    def available(self):
+        return log_info.has_log_file
+
+    @property
+    def page(self):
+        try:
+            return int(self.request.form.get('page', 1))
+        except ValueError:
+            return 1
 
-    def __call__(self):
-        return self.render()
+    @property
+    def size(self):
+        try:
+            return int(self.request.form.get('size', DEFAULT_SIZE))
+        except ValueError:
+            return DEFAULT_SIZE
 
-    def render(self):
-        return self.index()
+    @property
+    def query(self):
+        return self.request.form.get('query', '').strip()
 
-    def available(self):
-        return LOGFILE is not None
+    def _filtered_audit_log_lines(self):
+        """Generator for requested filtered log lines
+        """
+        query = self.query
+        for line in all_audit_log_lines():
+            if not query or query in line:
+                yield line
 
-    def tail(self, f, n=100):
-        """Return the last n lines of a file.
-        See: http://stackoverflow.com/a/280083/644075
+    def _batched_audit_log_lines(self):
+        """Generator for requested batch of the already filtered log lines
         """
-        assert n >= 0
-        pos, lines = n + 1, []
-        while len(lines) <= n:
-            try:
-                f.seek(-pos, 2)
-            except IOError:
-                f.seek(0)
+        start = (self.page - 1) * self.size
+        end = start + self.size
+        for pos, line in enumerate(self._filtered_audit_log_lines()):
+            if pos >= end:
                 break
-            finally:
-                lines = list(f)
-            pos *= 2
-        return lines[-n:]
+            if pos < start:
+                continue
+            yield line
 
-    @property
-    def _get_audit_log_files(self):
-        """Return a list with the current audit log and all of its
-        backups. The list is sorted with newer files first.
+    def build_audit_log(self):
+        """Triggers build and sets attributes for use in template
         """
-        # retrieve the list of audit log backup files
-        logfiles = sorted([
-            f for f in glob(LOGFILE + '.*') if '.lock' not in f
-        ], reverse=True)
-        # include active audit log as first element
-        logfiles.insert(0, LOGFILE)
-        return logfiles
+        lines = [l for l in self._batched_audit_log_lines()]
+        self.audit_log = '\n'.join(lines)
+        self.has_next = len(lines) == self.size
```

Comment:
this is not working: I see and I can use the `next` option even when there is no next page.

Reply:
fixed, I forgot some omit-tags in the templates. Since I can not look forward, in the rare case one has e.g. a 100 batch and exact 100 items are there, then the next link appears. Since I don’t want to load the next full batch b/c of efficiency reasons (large batch would take twice as long) I think we can live with it. There's always a solution, but imo its not worth the effort for such a minor glitch.